packages/plugin-ext/src/hosted/browser/plugin-worker.ts

```diff
@@ -25,22 +27,41 @@ export class PluginWorker {
     public readonly rpc: RPCProtocol;
 
     constructor() {
-        const emitter = new Emitter<string>();
-
         this.worker = new Worker(new URL('./worker/worker-main',
             // @ts-expect-error (TS1343)
             // We compile to CommonJS but `import.meta` is still available in the browser
             import.meta.url));
 
-        this.worker.onmessage = m => emitter.fire(m.data);
-        this.worker.onerror = e => console.error(e);
+        const onCloseEmitter = new Emitter<ChannelCloseEvent>();
+        const onMessageEmitter = new Emitter<MessageProvider>();
+        const onErrorEmitter = new Emitter<unknown>();
 
         this.rpc = new RPCProtocolImpl({
-            onMessage: emitter.event,
-            send: (m: string) => {
-                this.worker.postMessage(m);
-            }
+            close: () => {
+                onCloseEmitter.dispose();
+                onErrorEmitter.dispose();
+                onMessageEmitter.dispose();
+            },
+            getWriteBuffer: () => {
+                const writer = new Uint8ArrayWriteBuffer();
+                writer.onCommit(buffer => {
+                    this.worker.postMessage(buffer);
+                });
+                return writer;
+            },
+            onClose: onCloseEmitter.event,
+            onError: onErrorEmitter.event,
+            onMessage: onMessageEmitter.event
+        });
+
+        // eslint-disable-next-line arrow-body-style
+        this.worker.onmessage = buffer => onMessageEmitter.fire(() => {
+            return new Uint8ArrayReadBuffer(buffer.data);
         });
+
+        this.worker.onerror = e => onErrorEmitter.fire(e);
+
+        this.worker.onerror = e => console.error(e);
```

Comment:
One of them is probably not needed.

Reply:
Fixed with https://github.com/eclipse-theia/theia/commit/2b4016e8da50466f271bce2b2f808bf0ed371a10